bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/IExpandableNode.java

```diff
@@ -0,0 +1,62 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Advantest Europe GmbH and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ * 				Raghunandana Murthappa
+ * 				Andrey Loskutov (loskutov@gmx.de)
+ *******************************************************************************/
+package org.eclipse.jface.viewers;
+
+/**
+ * The expandable placeholder element to be used for viewer items that represent
+ * an expandable tree or table element.
+ * <p>
+ * The idea of {@link IExpandableNode} is to allow viewers to show only some
+ * subset of children that would otherwise all appear below parent element. The
+ * main purpose of this is to prevent UI freezes with viewers that can offer lot
+ * of elements but can't efficiently handle such amount using SWT.
+ * <p>
+ * The node consists of a parent element, list of all children of this parent
+ * and the offset to which child elements are supposed to be created and shown
+ * in the viewer.
+ *
+ * @noimplement This interface is not intended to be implemented by clients.
+ * @since 3.31
+ */
+public interface IExpandableNode {
```

Comment:
With @laeubi mentioning that virtual trees/tables work well without such node, and as it seems that this node introduces complexity for consumers or dealing with selection, we should simply just remove this class, and instead of dealing with this node, keep track of the position of the last visible item.

Reply:
The problem are the clients that extend `TreeViewer.isExpandable()` like https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/632/files#diff-b39aefe8a69251e08c379bd294eb6502b536e0b3c57e1d473dfa1e09ceddca9dR428

Of course TreeViewer should not be extended but of course everyone does that...

Without knowing that the node is "special" the extended code will not properly work.

Reply:
That's why I believe it would be much simpler from a consumer POV to not have this node at all.

Reply:
> That's why I believe it would be much simpler from a consumer POV to not have this node at all.

So you want show an element in a viewer but not expose this node to the viewer API's? 
This would require low level work in the viewer (at SWT level) and adoption of all the code in the viewer that retrieves SWT data in some way. Not sure if that is doable at all - hide the element from subclasses that actually is there.

Or do you have something completely different in mind? If yes, what?

Reply:
The extra node is needed to:
1) Give user a hint "there is more"
2) Give the user a chance to "Click me to see more".

How is that above is doable without an extra element in the tree? 
Do you mean some extra clickable decoration on the "last" child node or on the parent node or on both?

Reply:
> The problem are the clients that extend TreeViewer.isExpandable() [...]
> Without knowing that the node is "special" the extended code will not properly work.

In such case it is best to have a method `protected boolean isSpecialNodeX(Object o);` so extenders can know its the special type but you do not need to expose how this is implemented internally.

Reply:
`IExpandableNode` is now internal and provided 'isExpandableNode(Object element)' protected API.